standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java

```diff
@@ -515,6 +529,803 @@ public List<String> getMaterializedViewsForRewriting(String dbName) throws MetaE
+          params[paramIndex++] = partColumnPrivilege.getGrantorType();
+          params[paramIndex++] = partColumnGrantIdToPartId.get(entry.getKey());
+          params[paramIndex++] = partColumnPrivilege.getPrincipalName();
+          params[paramIndex++] = partColumnPrivilege.getPrincipalType();
+          params[paramIndex++] = partColumnPrivilege.getPrivilege();
+        }
+        try (QueryWrapper query = new QueryWrapper(pm.newQuery("javax.jdo.query.SQL", batchQueryText))) {
+          executeWithArray(query.getInnerQuery(), params, batchQueryText);
+        }
+      }
+    };
+    insertInBatch(PART_COL_PRIVS, columns, columnCount, row, rowCount, bec);
+  }
+
+  /**
+   * Add partitions in batch using direct SQL
+   * @param parts list of partitions
+   * @param partPrivilegesList list of partition privileges
+   * @param partColPrivilegesList list of partition column privileges
+   * @throws MetaException
+   */
+  public void addPartitions(List<MPartition> parts, List<List<MPartitionPrivilege>> partPrivilegesList,
+      List<List<MPartitionColumnPrivilege>> partColPrivilegesList) throws MetaException {
+    Map<Long, MSerDeInfo> serdeIdToSerDeInfo = new HashMap<>();
+    Map<Long, MColumnDescriptor> cdIdToColumnDescriptor = new HashMap<>();
+    Map<Long, MStorageDescriptor> sdIdToStorageDescriptor = new HashMap<>();
+    Map<Long, MPartition> partIdToPartition = new HashMap<>();
+    Map<Long, MPartitionPrivilege> partGrantIdToPrivilege = new HashMap<>();
+    Map<Long, MPartitionColumnPrivilege> partColumnGrantIdToPrivilege = new HashMap<>();
+    Map<Long, Long> sdIdToSerdeId = new HashMap<>();
+    Map<Long, Long> sdIdToCdId = new HashMap<>();
+    Map<Long, Long> partIdToSdId = new HashMap<>();
+    Map<Long, List<String>> stringListIdToStringList = new HashMap<>();
+    Map<Long, Long> stringListIdToSdId = new HashMap<>();
+    Map<Long, String> stringListIdToLocation = new HashMap<>();
+    Map<Long, Long> partGrantIdToPartId = new HashMap<>();
+    Map<Long, Long> partColumnGrantIdToPartId = new HashMap<>();
+    List<Long> stringListIds = new ArrayList<>();
+    int partitionsCount = parts.size();
+    for (int index = 0; index < partitionsCount; index++) {
+      MPartition part = parts.get(index);
+      Long serDeId = getDataStoreId(MSerDeInfo.class);
+      serdeIdToSerDeInfo.put(serDeId, part.getSd().getSerDeInfo());
+
+      Long cdId;
+      DatastoreId storeId = (DatastoreId) pm.getObjectId(part.getSd().getCD());
+      if (storeId == null) {
+        cdId = getDataStoreId(MColumnDescriptor.class);
+        cdIdToColumnDescriptor.put(cdId, part.getSd().getCD());
+      } else {
+        cdId = (Long) storeId.getKeyAsObject();
+      }
+
+      Long sdId = getDataStoreId(MStorageDescriptor.class);
+      sdIdToStorageDescriptor.put(sdId, part.getSd());
+      sdIdToSerdeId.put(sdId, serDeId);
+      sdIdToCdId.put(sdId, cdId);
+
+      Long partId = getDataStoreId(MPartition.class);
```

Comment:
I'm wondering if the unique id would be conflict when there are multiple Metastore instances in the warehouse.

Reply:
Datanucleus generate identity values unique across different JVMs. We have used datanucleus api to get the id.